### PR TITLE
Allows Composite custom search parameters

### PR DESCRIPTION
**Pull request by brendankowitz**
## Description
Allows Composite custom search parameters

## Related issues
Addresses #2501.

## Testing
Adds http file example

## FHIR Team Checklist
- [x] **Update the title** of the PR to be succinct and less than 50 characters
- [x] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [x] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [x] Tag the PR with Azure API for FHIR if this will release to the Azure API for FHIR managed service (CosmosDB or common code related to service)
- [x] Tag the PR with Azure Healthcare APIs if this will release to the Azure Healthcare APIs managed service (Sql server or common code related to service)
- [ ] CI is green before merge
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch

### CODE REVIEW

File: docs/rest/Data/CustomSearchPatient.json [hunk 1]
```diff
@@ -87,6 +87,7 @@
     "birthDate": "2016-01-15",
     "address": [
         {
+            "use": "usual",
```
Comment (rotodd) [inline docs/rest/Data/CustomSearchPatient.json:90]: It looks like we need to pick the address use value from [this enum](http://hl7.org/fhir/valueset-address-use.html). I noticed we get a 400 error when we post this patient:

`"Error occurred when parsing model: 'While building a POCO: Literal 'usual' is not a valid value for enumeration 'AddressUse' (at Resource.address[0].use[0])'."`

File: docs/rest/Data/CompositeCustomSearchParameter.json [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    "code" : "address-use-postcode",
+    "base" : ["Patient"],
+    "type" : "composite",
+    "expression" : "Patient",
+    "xpathUsage" : "normal",
+    "multipleOr" : false,
+    "component" : [{
+      "definition" : "http://hl7.org/fhir/SearchParameter/individual-address-use",
+      "expression" : "address.use"
+    },
```
Comment (rotodd) [inline docs/rest/Data/CompositeCustomSearchParameter.json:37]: Nit: this composite search parameter can't be parsed properly if we are running STU3 - maybe we could add a comment at the top of the `.http` test file to point people to using R4 or R5?

File: docs/rest/customSearchExample.http [hunk 3]
```diff
@@ -94,3 +101,9 @@ Authorization: Bearer {{bearer.response.body.access_token}}
 GET https://{{hostname}}/Patient?race=2028-9
 Authorization: Bearer {{bearer.response.body.access_token}}
 x-ms-use-partial-indices: true
+
+### A search query showing the Composite custom search parameter with Use/Postcode pair.
+# The usecase would be to evaluate multiple values together in chained search queries
+GET https://{{hostname}}/Patient?address-use-postcode=home$74047
+Authorization: Bearer {{bearer.response.body.access_token}}
+x-ms-use-partial-indices: true
```
Comment (rotodd) [inline docs/rest/customSearchExample.http:109]: Super nit: could we remove this header?

File: src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/SearchParameterOperations.cs [hunk 4]
```diff
@@ -65,6 +65,15 @@ public async Task AddSearchParameterAsync(ITypedElement searchParam, Cancellatio
...
                 var searchParameterInfo = new SearchParameterInfo(searchParameterWrapper);
+
+                if (searchParameterInfo.Component?.Any() == true)
+                {
+                    foreach (SearchParameterComponentInfo c in searchParameterInfo.Component)
+                    {
+                        c.ResolvedSearchParameter = _searchParameterDefinitionManager.GetSearchParameter(c.DefinitionUrl.OriginalString);
+                    }
+                }
+
```
Comment (rotodd) [inline src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/SearchParameterOperations.cs:76]: Does it make sense to add a custom search test for this scenario?

**Review by rotodd — APPROVED** [review]

File: docs/rest/Data/CompositeCustomSearchParameter.json (hunk 2 above)
Comment (brendankowitz) [reply docs/rest/Data/CompositeCustomSearchParameter.json:37]: Comment added in http file

File: src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/SearchParameterOperations.cs (hunk 4 above)
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/SearchParameterOperations.cs:76]: Added a unit test to ensure components get resolved

**Review by rotodd — APPROVED** [review]

**Pull request merged**